Fail segment load on gross PTS deviation

stevemayhew opened the pull request:
If the ExoPlayer encouters an unreported PTS discontinuity in the media that exceeds segment size the player will ultimately freeze forever (with one of the renderers starved for samples in the buffering state, even though LoadControl reports fully bufferd).

This change detects the issue and fails the segment load.

Note, other players (ffplay, Apple's AVPlayer, etc) 'recover' from these streams, at least playing the stream perphaps with A/V sync issues.  These kind of stream issues regulary occur in broadcast streams where affiliate perform ad insertion on the fly.

Issue: #7030

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/source/SampleQueue.java:
```diff
@@ -673,6 +676,23 @@ private synchronized void commitSample(
     int relativeEndIndex = getRelativeIndex(length);
+
+    // Check for obvious discontinuities in the content
+    if ( length > 0 && (sampleFlags & C.BUFFER_FLAG_KEY_FRAME) == 1) {
```
How is this change going to work with subtitle and metadata tracks, which are often quite sparse? A 10 second gap is not unexpected, and probably occurs quite regularly.

In general, I worry that this kind of change solves one problem (for media that's actually broken) whilst risking introducing new problems (for media that's actually valid). It's not even obviously wrong for a video track to have a >10 second gap between frames, for some more unusual cases. For example if a video track is being used for album artwork.

stevemayhew replied:
We have sample streams from MusicChoice which we support that we can try, but I'm not a fan of hard-coded values.  The goal is to detect intra-Extractor deviations, so we could check against and average value, this dos need more thought.

stevemayhew commented:
I agree with you, a hard and fast 10s is not a good number, and certainly
not for non video tracks.  You can see from the sample video (broadcast of
the US ProBowl) it’s poorly handled ad insertions often. One other thought
is catching in the `Extractor`. Doing nothing (ExoPlayer freezes then), is
not an option for our customers.  They simply blame it on ExoPlayer (“other
players work”) making it difficult for us to convince them the origin needs
to fix the stream metadata

Perhaps some average and check if a sample deviates more than a standard
deviation from it



On Sun, Mar 1, 2020 at 2:09 PM Oliver Woodman <notifications@github.com>
wrote:

> *@ojw28* commented on this pull request.
> ------------------------------
>
> In
> library/core/src/main/java/com/google/android/exoplayer2/source/SampleQueue.java
> <https://github.com/google/ExoPlayer/pull/7034#discussion_r386145292>:
>
> > @@ -673,6 +676,23 @@ private synchronized void commitSample(
>      largestQueuedTimestampUs = Math.max(largestQueuedTimestampUs, timeUs);
>
>      int relativeEndIndex = getRelativeIndex(length);
> +
> +    // Check for obvious discontinuities in the content
> +    if ( length > 0 && (sampleFlags & C.BUFFER_FLAG_KEY_FRAME) == 1) {
>
> How is this change going to work with subtitle and metadata tracks, which
> are often quite sparse? A 10 second gap is not unexpected, and probably
> occurs quite regularly.
>
> In general, I worry that this kind of change solves one problem (for media
> that's actually broken) whilst risking introducing new problems (for media
> that's actually valid). It's not even obviously wrong for a video track to
> have a >10 second gap between frames, for some more unusual cases. For
> example if a video track is being used for album artwork.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ExoPlayer/pull/7034?email_source=notifications&email_token=AADBF6B6WN2VQZNLSHQF3LLRFLMI5A5CNFSM4K64TA22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXPE4YY#pullrequestreview-366890595>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADBF6GGLFMLWXKCAOL2EWDRFLMI5ANCNFSM4K64TA2Q>
> .
>

ojw28 commented:
Presumably you can point to Safari losing A/V sync as a second player that's not correctly handling this type of stream... It seems somewhat hard to argue that it's ExoPlayer's fault when it's only "some other players work", and also when the streams are clearly in violation of the spec ;).

I agree we should try and improve the current behaviour, but we need to come up with a way of doing so that clearly and obviously doesn't negatively impact on any other case (and where that can be determined just by looking at the code change). So I think we need to think a bit about what the best way of doing that is.

stevemayhew committed:
Add HlsCheckedSampleQueue to check timstamp range

Add a SampleQueue subclass that checks the timestamp range of media samples queued to it and reports an exception on load if the timestamp is outside of spec bounds.
(Smashed to a single commit prior to rebase)

stevemayhew commented:
I've updated to match the design in the bug, we are running with this change back-ported to current release-v2 to test it.  See comments in the commits for details.

stevemayhew reviewed (commented):
I will squash out this commit (468c200) in the final pull branch, it reverts the changes to the SampleQueued.  Executing the check in the SampleQueue itself is not a good idea, as it has no context to the source time ranges.  See the added class `HlsCheckedSampleQueue`

stevemayhew commented on library/core/src/main/java/com/google/android/exoplayer2/source/SampleQueue.java:
```diff
@@ -480,7 +480,7 @@ public final void sampleData(ParsableByteArray buffer, int length) {
 
   @Override
-  public final void sampleMetadata(
+  public void sampleMetadata(
```
Allow subclass.

stevemayhew commented on library/core/src/main/java/com/google/android/exoplayer2/source/UnreportedDiscontinuityException.java:
```diff
@@ -11,16 +12,15 @@
    */
-  public UnreportedDiscontinuityException(long timesUs, int sourceId) {
+  public UnreportedDiscontinuityException(long timesUs, Uri uri) {
+    super("Unreported discontinuity timeMs: " + C.usToMs(timesUs) + " in URI: " + uri);
```
Pass the actual segment's URI, much more useful.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsCheckedSampleQueue.java:
```diff
@@ -0,0 +1,65 @@
+
+public class HlsCheckedSampleQueue extends SampleQueue {
+  private static final String TAG = "HlsCheckedSampleQueue";
+
```
I will add a detailed class comment when this is finished.

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1340,7 +1341,7 @@ private static DummyTrackOutput createDummyTrackOutput(int id, int type) {
   }
 
-  private static final class FormatAdjustingSampleQueue extends SampleQueue {
+  private static final class FormatAdjustingSampleQueue extends HlsCheckedSampleQueue {
```
Can we just rename this to `HlsSampleQueue` and have it do both jobs, rather than creating another layer in the class hierarchy?

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/source/UnreportedDiscontinuityException.java:
```diff
@@ -0,0 +1,26 @@
+ * caused by a discontinuity in a segment that was not split and reported by metadata in
+ * an HLS (EXT-X-DISCONTINUITY) or DASH stream.
+ */
+public class UnreportedDiscontinuityException extends RuntimeException {
```
IMO make it extend `IOException` so you don't have to catch and re-throw.

The naming is also a bit confusing, because the exception is by definition reporting the discontinuity :). Maybe call it `UnexpectedDiscontinuityException`.

Also make it final.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/source/UnreportedDiscontinuityException.java:
```diff
@@ -0,0 +1,26 @@
+package com.google.android.exoplayer2.source;
```
Please add the copyright header (copy from another file + update the year)

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/source/UnreportedDiscontinuityException.java:
```diff
@@ -0,0 +1,26 @@
+  /**
+   * Consturct the exception
+   *
+   * @param timesUs last timestamp before attempted commit of the deviant sample
```
It would be nice to capture both timestamps: `lastSampleTimeUs` and `nextSampleTimeUs`. And to pass the whole `DataSpec` rather than just the `Uri`.

(Note that some HLS streams will use the same `Uri` for every segment, so at least knowing the range is pretty essential)

stevemayhew replied:
Excellent idea.  Yes, `lastSampleTimeUs` will give us an idea how far into the issue occurs segment it is (`lastSampleTimeUs` - `lastSampleTimeUs`).

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsCheckedSampleQueue.java:
```diff
@@ -0,0 +1,65 @@
+  @Override
+  public void sampleMetadata(long timeUs, int flags, int size, int offset,  @Nullable CryptoData cryptoData) {
+    if (lowestTimeUs != C.TIME_UNSET && timeUs < lowestTimeUs && ! loggedFirst) {
+      Log.d(TAG, "sampleMetadata() - committed timeUs: " + timeUs + " is " + C.usToMs(lowestTimeUs - timeUs) + "ms less then segment start time.  chunk: " + chunk.dataSpec.uri);
```
Please can we remove all the logging.

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsCheckedSampleQueue.java:
```diff
@@ -0,0 +1,65 @@
+      Log.d(TAG, "sampleMetadata() - committed timeUs: " + timeUs + " is " + C.usToMs(lowestTimeUs - timeUs) + "ms less then segment start time.  chunk: " + chunk.dataSpec.uri);
+      loggedFirst = true;
+    }
+    if (lowestTimeUs != C.TIME_UNSET && timeUs < (lowestTimeUs - C.msToUs(50_000))) {
```
Where is the magic 50 coming from? Shouldn't we just subtract `tolerance` from lowestTimeUs` in `setCurrentLoadingChunk`?

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsCheckedSampleQueue.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  void setCurrentLoadingChunk(HlsMediaChunk chunk) {
+    double tolerance = (chunk.endTimeUs - chunk.startTimeUs) * 0.1;
```
Is there any reason not to just use a constant (and quite large) value here, like 5 seconds?

stevemayhew committed:
Update with first round of code review suggestions

Update with code review suggestions:
1. rename to HlsCheckedSampleQueue to HlsSampleQueue and combine with format adjusting class.
2. Copywrite in added classes
3. Capture additional items useful for recovery and reporting in the exception class
4. Remove extraneous logging
5. eliminate magic number (50 seconds) and use percentage of duration

WIP:
 - eliminate null check for chunk (chunkless prepare starts load before sampleQueue are created)
 - work out recovery strategy

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleQueue.java:
```diff
@@ -0,0 +1,159 @@
+  public void sampleMetadata(long timeUs, int flags, int size, int offset,  @Nullable CryptoData cryptoData) {
+    // TODO - chunkless prepare, sampleQueue list is not yet initialized for first chunk
+//    Assertions.checkNotNull(chunk, "sampleMetadata without a MediaChunk?");
+    if (chunk == null) {
```
Had to replace this with null check, the initial segment loads are started before `HlsSampleStreamWrapper#createSampleQueue()` is called so need a way to get the loading chunk in that method and set it in the `HlsSampleQueue`

Or we ignore checking the first segment loads.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1343,82 +1343,6 @@ private static DummyTrackOutput createDummyTrackOutput(int id, int type) {
 
-  private static final class FormatAdjustingSampleQueue extends HlsCheckedSampleQueue {
-
-    private final Map<String, DrmInitData> overridingDrmInitData;
```
Combined into `HlsSampleQueue`

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java:
```diff
@@ -379,10 +379,9 @@ private void feedDataToExtractor(
+      } catch (UnexpectedDiscontinuityException e) {
+        Log.d(TAG, "UnexpectedDiscontinuityException - recovering by discarding balance of segment", e);
+        throw new IOException("load aborted for segment - " + e.dataSpec + " unexpected discontinuity", e);
       } finally {
```
`UnexpectedDiscontinuityException` needs to stay a `RuntimeException` until signatures in the class hierarchy for `TrackOutput#sampleMetadata()` are changed to include a declared exception.  Best to keep this in a separate change to avoid having to merge/rebase lots of files while this branch is open.

ojw28 replied:
Can you just make it an `IOException`, to avoid needing to catch and re-throw?

stevemayhew replied:
I still need to catch in order to get to the timestamp adjuster (and/or add metadata to the segment to discard the time period and recover).  I can probably do this on my own fork if no one else would want it (my preference always is ExoPlayer proper to take the changes, note if by default we fail playback anything done here to facilitate recovery will not matter).

This latest checkin (4701a6b39081cccaab202effb6cf59de6d4510b2) make the exception extend IOException (much nicer, I agree, then a runtime exception).   It took a while to make and run the tests, so hopefully I'll not have to rebase it much (Most of it is of course pretty rote to review and in the extractor library)

stevemayhew commented:
@ojw28 I rebased to dev-v2 and addressed (hopefully) all your review comments,  feel free to be picky, I'm happy to accommodate.

Currently I'm playing broadcasts streams of major network from multiple affiliates and multiple origin servers we support to make sure they playback normally.  The pro bowl sample I attached to the bug fails with a playback error rather then hanging using the current state of this branch.

I'm working on the recovery code now (looks like this will involve reseting the timestamp adjustment and discarding the balance of the segment).  Also welcome more discussion as to if this is a load error or some other event.  My personal feeling is treat it like any other load error, so:

1. the `HlsMediaChunk#feedDataToExtractor()` would setup any precursor to possible recovery (timestamp adjustment, etc.)
1. catch exception, log and re-throw so error handler would get it
1. if `LoadErrorAction DONT_RETRY` is returned, playback continues with balance of segment discarded.

The player UI's all have the option of implementing `LoadErrorHandlingPolicy` and choosing how to deal with this.

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleQueue.java:
```diff
@@ -0,0 +1,159 @@
+ * Apple's mediastreamvalidator.
+ *
+ */
+public class HlsSampleQueue extends SampleQueue {
```
Can this stay where `FormatAdjustingSampleQueue` was (i.e., as an inner class)? We definitely don't want it to be part of the ExoPlayer library's actual API surface, and I don't think enough is really being added to justify splitting it into its own file.

stevemayhew replied:
IMHO, breaking up big classes is a good thing (less conflicted merges, etc), I can make it package private and exclude it from the JavaDoc, but I do think it is a good idea to make it a standalone class.  Pretty much equivalent from the VM's view to a static inner class.

If I can exclude it from the package classes are you ok with this?  Otherwise, sure I'll move it to an inner class.

ojw28 replied:
It's not a great idea to split it out whilst simultaneously modifying it, because it makes the change harder to review (or look at later to see what changed). If it's to be split out, that should IMO be a separate change after this one, that's a pure no-op move rather than one that also changes behavior.

stevemayhew replied:
Ah. Yes, I agree 100%, I"ll move it back and reduce the diffs.

stevemayhew replied:
@ojw28 In this commit ccf2ba3e1 I moved the class back as an inner class and tried to remove the diffs.   Once we finish the pull up and agree it is merge ready I will rebase out the intermediate states to further reduce the changes to `HlsSampleStreamWrapper` so the y are clear.

Hope this helps.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/source/UnexpectedDiscontinuityException.java:
```diff
@@ -0,0 +1,68 @@
+   */
+  public final DataSpec dataSpec;
+
+  /** The timeUs that the source of the samples starts (from HLS metadata)
```
The Javadoc for this class seems confused about whether it's an HLS specific exception or not. Can we either move it into the HLS package and make it HLS specific, or remove specific references to HLS from the documentation?

stevemayhew replied:
The same problem could exist in DASH (at least as I understand it), the origin must split the timeline into multiple periods if the timestamps wrap or jump (https://dashif-documents.azurewebsites.net/Guidelines-TimingModel/master/Guidelines-TimingModel.html#timing-sampletimeline) I don't have a sample DASH playlist that breaks this rule.

ojw28 reviewed (commented):
Looking good!

stevemayhew committed:
Update comments for DASH

Update the comment to indicate the DASH equivalent use case for the exception.

stevemayhew committed:
Move HlsSampleQueue to be static inner-class

Backout making the original `FormatAdjustingSampleQueue` an outer class and combining the new timestamp checking logic.  This way the diff from `HlsSampleStreamWrapper` to dev-v2 are easier to see.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1337,26 +1340,57 @@ private static boolean formatsMatch(Format manifestFormat, Format sampleFormat)
     return true;
   }
 
-  private static DummyTrackOutput createDummyTrackOutput(int id, int type) {
-    Log.w(TAG, "Unmapped track with id " + id + " of type " + type);
-    return new DummyTrackOutput();
-  }
```
Patch diff moved this to after the original `FormatAdjustingSampleQueue` class.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1337,26 +1340,57 @@ private static boolean formatsMatch(Format manifestFormat, Format sampleFormat)
+    private long lowestTimeUs = C.TIME_UNSET;
+    private long highestTimeUs = C.TIME_UNSET;
 
-  private static final class FormatAdjustingSampleQueue extends SampleQueue {
```
Rename `FormatAdjustingSampleQueue` to `HlsSampleQueue` and add checks for timestamp discontinuity.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1337,26 +1340,57 @@ private static boolean formatsMatch(Format manifestFormat, Format sampleFormat)
+            DrmSessionManager<?> drmSessionManager,
+            MediaSourceEventDispatcher eventDispatcher,
+            Map<String, DrmInitData> overridingDrmInitData) {
       super(allocator, drmSessionManager, eventDispatcher);
```
spacing change obfuscates the simple rename.

stevemayhew commented:
Thanks @ojw28 for reviewing this, I know it has become a pretty large change.  This will make ExoPlayer less vulnerable to infinite stalls.  

Our team deals with lots of live streams from a growing variety of origin servers, each with their own "creative" solution to the discontinuity problem.  The more we can at least detect and better recover or at least report these errors the better our user experience is.

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1416,6 +1450,32 @@ private Metadata getAdjustedMetadata(@Nullable Metadata metadata) {
+        @Nullable CryptoData cryptoData) throws IOException {
+      // TODO - chunkless prepare, sampleQueue list is not yet initialized for first chunk
+  //    Assertions.checkNotNull(chunk, "sampleMetadata without a MediaChunk?");
+      if (chunk == null) {
```
I don't think I understand when this can happen, and the TODO above didn't help me figure it out. Could you elaborate?

stevemayhew replied:
The first segments to load when it is not [chunk-less prepare mode](https://medium.com/google-exoplayer/faster-hls-preparation-f6611aa15ea6).  This is because until the prepare is complete the `SampleStream` is not added to the `HlsSampleStreamWrapper.sampleQueues[]` array and the `HlsSampleStreamWrapper.init()` method does not get called until the load has already started.  So samples can be committed before the current loading `MediaChunk` is set.  

I'm working out how architecturally to make this happen, so the TODO is for me.

ojw28 replied:
Sorry, I still don't understand this. It looks to me like `HlsSampleStreamWrapper.init()` is always called before a chunk starts to load. Can you actually reproduce this not happening, and if so, how?

It's also unclear whether this is about chunk-less or non-chunk-less preparation. The comment says one thing, but the conversation says another. Please clarify; thanks!

ojw28 replied:
I see how this happens now.

ojw28 replied:
I fixed this as part of the merge.

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -1337,26 +1340,57 @@ private static boolean formatsMatch(Format manifestFormat, Format sampleFormat)
 
-    public void setDrmInitData(@Nullable DrmInitData drmInitData) {
+    void setCurrentLoadingChunk(HlsMediaChunk chunk) {
+      double tolerance = (chunk.endTimeUs - chunk.startTimeUs) * MAX_TIMESTAMP_DEVIATION_PERCENTAGE;
```
For small segments this can end up allowing quite a small tolerance. I wonder whether it's worth adding a minimum tolerance (e.g., 5 seconds) as well, and `Math.max`-ing the two? I kind of wonder whether just a fixed tolerance (e.g., 10 seconds) might be easier to reason above, since I'm not sure you'd necessarily want to allow a really large deviation for long segments either.

stevemayhew replied:
I agree.   I have increased this in my own branch and testing to a larger value,  this is particularly evident for i-Frame only which can be of extremely short times.  The `Math.max()` is a great way to do this.

stevemayhew replied:
See commit 4859207 changed to make max the tolerance against a min possible value.

ojw28 reviewed (commented):
Sorry for the delay. See inline comments.

I didn't realize having the exception extend IOException would require so many files to be touched! I'll probably put it back to a runtime exception as you had it (probably `IllegalStateException`) when I merge this, to avoid that. No need to do it on your side though.

stevemayhew commented:
> Sorry for the delay. See inline comments.
> 
> I didn't realize having the exception extend IOException would require so many files to be touched! I'll probably put it back to a runtime exception as you had it (probably `IllegalStateException`) when I merge this, to avoid that. No need to do it on your side though.

Not a problem,  I pretty much guessed you would feel this way... So I purposely made the commit so I can back it out (branched before) so I'll remove it and change to `IllegalStateException`

ojw28 commented:
> Not a problem, I pretty much guessed you would feel this way... So I purposely made the commit so I can back it out (branched before) so I'll remove it and change to IllegalStateException

Thanks. Please could you push this, and we'll look at getting this merged.

stevemayhew committed:
Move back to `UnexpectedDiscontinuityException` extending a `RuntimeException`

This avoids massive changes to method signatures to add throws.  Also, took suggestion to make it an `IllegalStateException`.

Move the catch outside of the finally that sets `nextLoadPosition` (this allows for possible recovery by reseting the `Extractor` and `TimestampAdjuster`).

Lastly, took the suggestion to make a minimum value for the tolerance (especially usefull for very short i-Frame only segments).

stevemayhew commented:
@ojw28 the force push eliminated the commit 4701a6b (changing from `RuntimeException`) so all the methods do not need signature change.  Also, took your suggestions in the latest commit 4859207.  I've updated our internal release branch to match these changes to test them out.

ojw28 commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java:
```diff
@@ -379,6 +382,9 @@ private void feedDataToExtractor(
         nextLoadPosition = (int) (input.getPosition() - dataSpec.position);
       }
+    } catch (UnexpectedDiscontinuityException e) {
+      Log.d(TAG, "UnexpectedDiscontinuityException - recovering by discarding balance of segment", e);
```
What does "discarding balance of segment" mean? Does this actually handle recovery in its current form? I'm not sure how it does, if so?

stevemayhew replied:
The balance of the chunk's samples simply follow the new PTS sequence, so without a fresh timestamp adjuster, they are useless.

With the last checkin the exception is reported as a playback error  (normal error handler in `Loader` wraps it in an `UnexpectedLoaderException` so `DefaultLoadErrorHandlingPolicy.getRetryDelayMsFor()` will see an `UnexpectedLoaderException` then `Loader.DONT_RETRY_FATAL` is returned so the main playback thread is notified)

I'm working on a partial recovery, which would involve skipping at least to the next chunk, like a seek outside of the `SampleQueue`  range (reseting timestampAdjuster).  This works manually if you add code to 'ignore' the error, like:

````java
  @Override
  public LoadErrorAction onLoadError(
      Chunk loadable,
      long elapsedRealtimeMs,
      long loadDurationMs,
      IOException error,
      int errorCount) {
...
    if (error.getCause() instanceof UnexpectedDiscontinuityException) {
      loadErrorAction = Loader.DONT_RETRY;    // Ignore the balance of this chunk
    }

    eventDispatcher.loadError(
        loadable.dataSpec,
...
    return loadErrorAction;
  }
````

ojw28 commented:
I merged a variant of most of this change, in https://github.com/google/ExoPlayer/commit/8991586c3b9b17a71214dfdeeaf679c66565a591. Note that:

* I found a way to make the exception extend `IOException` and not have to modify all the other files (it's super hacky, but still marginally better than a `RuntimeException` that then needs to be wrapped to re-throw.
* Actual throwing of the exception is commented out with a TODO. If you could send another pull request that enables it and also enables recovery at the same time, I think that would be ideal.
* I scoped things so that this is limited to HLS for now.

So I think next steps are to figure out how to make recovery work. Thanks!

stevemayhew commented:
Thanks, this is great.  I can pull this into a new branch from `dev-v2` then and add the recovery.   My thinking is to do it in the demo player in an `LoadErrorPolicy` (by skipping past the bad content, as the seek resets the timestamp adjustment)

What does this mean?
> I scoped things so that this is limited to HLS for now.

Anything special I need to do to include the change in the scope in building the demo?

ojw28 commented:
I was hoping it would be possible just to do a sensible recovery inside the HLS module, rather than having to get things like `LoadErrorPolicy` involved.

> What does this mean?

It just means that the exception was moved to the HLS module and made package private there. Which makes sense if recovery can be done as per my comment above. If it needs to happen somewhere else the exception might need moving and making public.

stevemayhew commented:
@ojw28 maybe..  I think you'll find the same issue is possible in live playback in DASH, as I mentioned in the pull request.

ojw28 commented:
I don't want to encourage bad serving implementations by handling spec violations that aren't already proven to be deeply rooted in existing solutions. I don't think we've seen evidence that this is significant problem in DASH solutions, and so I don't think we have evidence that this will ever be required for DASH.

I also suspect that it's much harder to splice random things together in an FMP4 container and still produce a segment that's vaguely close to being valid, so it may be that this makes it less likely for DASH solutions to end up with this kind of problem.
